slicing across cubes

abit2 opened the pull request:
slicing across cubes. @DanRyanIrish

pep8speaks commented:
Hello @abit2! Thanks for updating the PR.

 - In the file [`sunpycube/cube/cube_utils.py`](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py), following are the PEP8 issues :

> [Line 527:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L527): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 536:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L536): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)
> [Line 551:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L551): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 578:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L578): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 581:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L581): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 584:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L584): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 597:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L597): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 607:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L607): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 617:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L617): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 620:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L620): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 621:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L621): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 630:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L630): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 633:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L633): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 634:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L634): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 647:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L647): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 667:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L667): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 668:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L668): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 670:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L670): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 674:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L674): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 678:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L678): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 680:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L680): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 683:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L683): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 690:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L690): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 698:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L698): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 700:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L700): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 701:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/cube_utils.py#L701): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 79 characters)

 - In the file [`sunpycube/cube/datacube.py`](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py), following are the PEP8 issues :

> [Line 82:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L82): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 102:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L102): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 135:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L135): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 152:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L152): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 170:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L170): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 171:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L171): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 173:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L173): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 174:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L174): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 175:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L175): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 551:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L551): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 552:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L552): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 553:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L553): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 554:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L554): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 611:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L611): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 617:4](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L617): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 637:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/datacube.py#L637): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)

 - In the file [`sunpycube/cube/tests/test_cube_utils.py`](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py), following are the PEP8 issues :

> [Line 9:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L9): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)
> [Line 10:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L10): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 79 characters)
> [Line 11:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L11): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)
> [Line 16:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L16): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 79 characters)
> [Line 17:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L17): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 18:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L18): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)
> [Line 131:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L131): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 132:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L132): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 133:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L133): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)
> [Line 134:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L134): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)
> [Line 142:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L142): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 79 characters)
> [Line 146:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L146): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 79 characters)
> [Line 150:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L150): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 79 characters)
> [Line 153:34](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cube_utils.py#L153): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

 - In the file [`sunpycube/cube/tests/test_cubesequence.py`](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cubesequence.py), following are the PEP8 issues :

> [Line 22:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cubesequence.py#L22): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 23:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cubesequence.py#L23): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 79 characters)
> [Line 24:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cubesequence.py#L24): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 27:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cubesequence.py#L27): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 79 characters)
> [Line 28:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cubesequence.py#L28): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 29:80](https://github.com/sunpy/cube/blob/6df7518af7c9d3770a04e4f47febcd06e00a1b6e/sunpycube/cube/tests/test_cubesequence.py#L29): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)



##### Comment last updated on July 13, 2017 at 19:30 Hours UTC

abit2 commented:
corrected the failing test pixelize_slice (no segmentation fault :P). Added test and made some changes/added lines in code. Made comments in code this time. :) @DanRyanIrish

DanRyanIrish commented on sunpycube/cube/cube_utils.py:
```diff
@@ -613,10 +633,10 @@ def index_sequence_as_cube(cubesequence, item):
+        # if sequence_index.start != sequence_index.stop:
+        #     raise ValueError("Slicing across multiple cubes not yet supported.")
+        # else:
+        #     sequence_index = sequence_index.start
```
Delete these commented out lines as slicing across cubes should now be supported.

DanRyanIrish commented on sunpycube/cube/cube_utils.py:
```diff
@@ -519,16 +517,38 @@ def get_cube_from_sequence(cubesequence, item):
-        meta = cubesequence.meta
-        result = cubesequence._new_instance(data, meta)
-
+        data = cubesequence.data[item.start:item.stop]
```
Is there a reason you don't you just index with the slice object itself?  That way a user could use the slice step to get e.g. every 2nd cube if they want.

DanRyanIrish replied:
I see you indexed with just the slice below.  Is there a reason you haven't done that here too?

abit2 replied:
no it is my mistake will be changing this :)

DanRyanIrish commented on sunpycube/cube/cube_utils.py:
```diff
@@ -636,19 +656,53 @@ def _convert_cube_like_index_to_sequence_indices(cube_like_index, cumul_cube_len
-        sequence_index += 1
+        # if the cube is out of range then return the last index
+        if cube_like_index > cumul_cube_lengths[-1] - 1:
+            cube_index = cumul_cube_lengths[0] - 1
```
If the user enters an exposure number which is out of range of the number of exposures in the CubeSequence, I think an error should be raised rather.

abit2 replied:
```python
In [4]: a = np.array([1,2,3,4,5])

In [5]: a[0:100]
Out[5]: array([1, 2, 3, 4, 5])

```

@DanRyanIrish array behaves in this way so thought it should be same.

DanRyanIrish replied:
Wow!  I did not realise that!  In that case you've done the right thing :)

DanRyanIrish commented on sunpycube/cube/datacube.py:
```diff
@@ -554,20 +555,16 @@ class CubeSequence(object):
+        except AttributeError as err:
+            warnings.warn("AttributeError " + str(err))
+        else:
+            self.shape = tuple([len(data_list)])
```
Is the purpose of this try...except statement to make CubeSequence work for non-Cubes/arrays?

abit2 replied:
Yes!! lightcurve does not have shape attribute

DanRyanIrish commented on sunpycube/cube/tests/test_cube_utils.py:
```diff
@@ -115,13 +115,39 @@ def test_convert_slice():
+    (cu._convert_cube_like_index_to_sequence_indices(50, np.array([8, 16, 24, 32])), (3, 7)),
+])
+def test_convert_cube_like_index_to_sequence_indices(test_input, expected):
+    assert test_input == expected
```
Does this and other tests also check whether it errors in the right way?

DanRyanIrish commented on sunpycube/cube/tests/test_cube_utils.py:
```diff
@@ -115,13 +115,39 @@ def test_convert_slice():
+
+
+@pytest.mark.parametrize("test_input,expected", [
+    (cu._convert_cube_like_slice_to_sequence_slices(slice(2, 5),
```
Isn't this a circular test?  Using the function to give the answer that you want to test? If so, wont that mean it wont fail even if the function is wrong?  Would it be better better to have explicit expected inputs and outputs?

DanRyanIrish commented on sunpycube/cube/tests/test_cube_utils.py:
```diff
@@ -115,13 +115,39 @@ def test_convert_slice():
+    (cu._convert_cube_like_slice_to_sequence_slices(slice(5, 15), np.array(
+        [8, 16, 24, 32])), (slice(0, 2), [slice(5, 8), slice(0, 7)])),
+    (cu._convert_cube_like_slice_to_sequence_slices(slice(5, 16),
+                                                    np.array([8, 16, 24, 32])), (slice(0, 2), [slice(5, 8)])),
```
Should the end of this be ```[slice(5, 8), slice(0, 8)]``` ?

DanRyanIrish commented on sunpycube/cube/tests/test_cube_utils.py:
```diff
@@ -115,13 +115,39 @@ def test_convert_slice():
+    (cu._convert_cube_like_slice_to_sequence_slices(slice(5, 23), np.array(
+        [8, 16, 24, 32])), (slice(0, 3), [slice(5, 8), slice(0, 7)])),
+    (cu._convert_cube_like_slice_to_sequence_slices(slice(5, 100),
+                                                    np.array([8, 16, 24, 32])), (slice(0, 4), [slice(5, 8)])),
```
Should the output of this be ```(slice(0, 6), [slice(5, 8), slice(0,4)])```?

abit2 replied:
No the tuple[0] is the sequence_slice which will be from (0, 4) and the tuple[1] is the list which consist of the slices of 1st and last cube which if it needs to be sliced.

DanRyanIrish replied:
OK.  It's good to have that confirmed.  But here, you want to get the 5th to the 99th (inclusive) exposures, right?  There are 16 raster positions in each cube, right?  So the first exposure of the 7th cube brings you to the 96th exposure.  (This is all 0-based counting.)  So to get to the 100th exposure you need slice(0,4) of the 7th cube.  So in fact you need ```(slice(0, 7), [slice(5, 8), slice(0,4)])```?

DanRyanIrish replied:
Fixed.

DanRyanIrish commented on sunpycube/cube/tests/test_cubesequence.py:
```diff
@@ -58,6 +60,22 @@ def test_slice_first_index_sequence(test_input, expected):
+
+
+@pytest.mark.parametrize("test_input,expected", [
+    (seq.index_as_cube(0), Spectrum),
```
Should this not also be a GenericMap?
